[RFC]oldtests: win: run on cmd.exe

janlazo opened the pull request:
oldtests are failing on Appveyor but the builds pass because the error code is ignored.
Make them pass and check the error code of make.exe

justinmk commented:
Let's do this in small steps: the greater of "1 test file or 50-100 line changes" per PR. That will let us move faster.

So "check the error code of make.exe" will wait until the end.

janlazo commented:
Most of the noise in oldtests logs for the Appveyor builds is because `unix.vim` wasn't sourced.
https://github.com/neovim/neovim/commit/4eba134a551161384428229fbd80721eb49b5a09#diff-b3c6ad6680a25a1b42095879e3a87104R168 appends`-u NONE` after RUN_VIMTEST.

This caused the new-style Vim tests to run on cmd.exe instead of sh.

justinmk commented on src/nvim/testdir/Makefile:
```diff
@@ -230,4 +230,4 @@ newtestssilent: $(NEW_TESTS)
 	@echo "[OLDTEST] Running" $*
 	@mkdir -p $(TMPDIR)
-	@/bin/sh runnvim.sh $(ROOT) $(NVIM_PRG) $* $(RUN_VIMTEST) -u NONE -S runtest.vim $*.vim
+	@/bin/sh runnvim.sh $(ROOT) $(NVIM_PRG) $* $(RUN_VIMTEST) -S runtest.vim $*.vim
```
maybe @ZyX-I meant for it to be capital `-U NONE`, like https://github.com/janlazo/neovim/blob/0dad25f010db326e3689b136e8385fd03d781b0e/src/nvim/testdir/Makefile#L175 ?

janlazo commented:
@justinmk Do you know why `sourcing unix.vim` causes the oldtests to fail in Travis, specifically the filetype detection test?

justinmk commented:
@janlazo maybe because of `set shellslash`?

janlazo commented:
Where is error code 2816 coming from?

justinmk commented on src/nvim/testdir/test_autocmd.vim:
```diff
@@ -816,8 +816,8 @@ func Test_Cmdline()
-  call assert_equal('/', g:entered)
-  call assert_equal('/', g:left)
+  call assert_equal(has('win32') ? 'C:/' : '/', g:entered)
+  call assert_equal(has('win32') ? 'C:/' : '/', g:left)
```
This isn't needed in upstream Vim: https://github.com/vim/vim/blob/e7499ddc33508d3d341e96f84a0e7b95b2d6927c/src/testdir/test_autocmd.vim#L816

When changing the oldtests, we need to be careful not to add a bunch of tiny changes except where absolutely unavoidable. Otherwise it makes merging pain in the ass.

And we need to ask why the result is different. Is this an intentional behavior change in `expand('<afile>')` that we plan to keep? If not, there's no reason to make the test pass.

janlazo replied:
Are they the same internally but Vim hides the network drive for `<afile>`?

justinmk commented on src/nvim/testdir/test_autocmd.vim:
```diff
@@ -372,7 +372,7 @@ func Test_autocmd_bufwipe_in_SessLoadPost()
         \ ]
   call writefile(content, 'Xvimrc')
-  call system(v:progpath. ' --headless -i NONE -u Xvimrc --noplugins -S Session.vim -c cq')
+  call system(expand(v:progpath). ' --headless -i NONE -u Xvimrc --noplugins -S Session.vim -c cq')
```
Already changes this line to add `--headless` so adding another change is fine.

Though I would like to eventually change Nvim so that `-es` works without `--headless`.

janlazo replied:
Sure for `set shellslash` and `set shell=sh` (or bash, wsl).

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -8552,7 +8552,12 @@ eval_vars (
+        } else {
+          result = (char_u *)FullName_save((char *)autocmd_fname, FALSE);
+        }
         xfree(autocmd_fname);
```
@bfredl does this change make sense to you?

@janlazo  BTW this block was changed on master, needs a rebase.

janlazo commented:
test_autocmd.vim has 1 failure only so I guess it worked?

janlazo commented:
Running nvim on sh/bash with `TERM=cygwin` results in segfault (negative exit code)  which means any tests that require nvim running on sh won't work if unsetting `TERM` doesn't work.

janlazo committed:
oldtests: newstyle: win: reset shell to cmd.exe

Unset $SHELL so that child nvim use cmd.exe as default shell.

Unset $TERM so that child nvim don't segfault with negative exit code.
sh/bash use TERM=cygwin by default if it is unset.
mintty sets TERM to xterm.

janlazo committed:
ci/win: prefer msys find.exe

runnvim.sh depends on it for file glob patterns.

janlazo committed:
oldtests: use expand() to fix pathsep

Fix failing oldtests because of noshellslash.

justinmk commented:
https://github.com/neovim/neovim/pull/7660 should avoid the segfault IIRC (though the UI isn't going to work)

janlazo commented:
test_profile.vim still has failures because of `SHELL=sh`, causing the segfault on child nvim. test_profile.vim passes on my Win 8.1 machine with the current shell defaults if I source the file directly instead of running make.

janlazo committed:
oldtests: win: cmd.exe echo has trailing CR

janlazo committed:
oldtests: win: shellslash for cdpath,tags,:find

janlazo commented:
`:find` requires `shellslash` for the first suggestion to appear but its based on the current directory, not `path`. With `noshellslash`, the suggestions are off-by-one (2nd becomes 1st, 3rd becomes 2nd, ...). Why is that?

janlazo commented:
`help sp?it` opens `split()`, not `:split` on Windows. It's the same on Vim.

janlazo committed:
oldtests: win: help sp?it opens split()

Vim 8 and Neovim 0.2.3 open the same target on Windows.

janlazo commented:
test_increment_dbcs.vim and test_makeencoding.vim are failing on MSVC, not MINGW.
@b-r-o-c-k Any ideas?

janlazo committed:
oldtests: check +iconv for makeencoding

MSVC builds don't include it.
Test utf8 and latin only if +iconv is missing.

b-r-o-c-k commented:
@janlazo No, the MSVC build doesn't use iconv, although I could add it to the bundled dependencies if desired. I tried running test_increment_dbcs.vim manually with both the MSVC and MinGW builds and they both fail with `Expected ['ŽR2'] but got ['ŽS1']`. I also checked the increment/decrement function `do_addsub` and I don't think it uses iconv.

bfredl commented on src/nvim/ex_docmd.c:
```diff
@@ -8544,7 +8544,14 @@ eval_vars (
+            || has_event(EVENT_CMDLINELEAVE)
+            || has_event(EVENT_CMDWINENTER)
+            || has_event(EVENT_CMDWINLEAVE)) {
+          result = (char_u *)tv_dict_get_string(get_vim_var_dict(VV_EVENT), "cmdtype", true);
```
Doesn't `result = xstrdup(autocmd_fname)` work? All these four autocommands should have set the same value to `fname`.

bfredl replied:
Also this if block changes unconditionally whenever any such autocommand is _defined_, not whether the actual _invoked_ event was one of these for. It doesn't make sense to me.

janlazo replied:
@bfredl How can I check the currently invoked event?

janlazo commented:
I don't know how to fix the `<afile>` issue so I'll revert the change. Do I have to submit any changes here to vim-dev (such as `expand`) since it runs on cmd.exe, not sh?

justinmk commented on src/nvim/testdir/runtest.vim:
```diff
@@ -42,6 +42,14 @@ endif
+  let $TERM = ''
+  let &shell = empty($COMSPEC) ? exepath('cmd.exe') : $COMSPEC
+  set shellcmdflag=/s/c shellxquote=\" shellredir=>%s\ 2>&1
+endif
```
should probably go in setup.vim

justinmk commented on src/nvim/testdir/test_system.vim:
```diff
@@ -12,7 +12,9 @@ function! Test_System()
-  if out != ["123 \r"]
+  if &shell =~# 'cmd.exe$'
+    call assert_equal(['123\r'], out)
+  elseif out != ["123 \r"]
```
why not just `else`?

janlazo replied:
The deleted line didn't include `has('win32')` so I thought it may happen in other OS.

justinmk commented on src/nvim/testdir/test_makeencoding.vim:
```diff
@@ -13,12 +13,19 @@ endif
 
-let s:message_tbl = {
+if has('iconv')
+  let s:message_tbl = {
```
when is iconv not available? The MSVC builds have it now.

janlazo replied:
https://github.com/neovim/neovim/pull/8160#issuecomment-377104244
The MSVC logs show that the iconv is not found, `-iconv` in Features.
In functional tests, I get this

```
  [----------] Running tests from C:/projects/neovim/test/functional\legacy\083_tag_search_with_file_encoding_spec.lua
  [ RUN      ] tag search with !_TAG_FILE_ENCODING skipped (missing iconv): 0.00 ms SKIP
  [----------] 1 test from C:/projects/neovim/test/functional\legacy\083_tag_search_with_file_encoding_spec.lua (47.00 ms total)
```

justinmk replied:
Was this rebased recently?

janlazo replied:
Rebasing did nothing. The MSVC build log still report that iconv.h is not found.

justinmk replied:
My mistake, I was thinking of libintl, not iconv.

justinmk commented:
> Do I have to submit any changes here to vim-dev (such as expand) since it runs on cmd.exe, not sh?

If you think they are applicable to Vim that would be helpful. But I would guess they're not applicable, else the tests wouldn't pass for Vim on appveyor ...

janlazo commented:
`:h getftype`
```
Note that a type such as "link" will only be returned on
systems that support it.  On some systems only "dir" and
"file" are returned.  On MS-Windows a symbolic link to a
directory returns "dir" instead of "link".
```

Why does Vim have this behaviour? Neovim returns `link` type in the symlink test as expected but Vim returns `dir` type.

justinmk commented:
> Neovim returns link type in the symlink test as expected but Vim returns dir type.

Let's stick with our behavior, it's ok to change the test .

janlazo committed:
win: getftype(symlink) returns 'link'

Vim doesn't detect symlinks correctly so stick with Neovim's behaviour.

janlazo commented:
```
From test_stat.vim:
Found errors in Test_existent_directory():
function RunTheTest[35]..Test_existent_directory line 5: Expected 'rwx' but got 'rw-'
```
Is this an upstream bug with libuv's us_fs_stat()?

janlazo commented:
Does neovim check if directory path in `dir` is valid such that it can't create a swapfile there and report `E303`?
If I use an actual restricted directory for `Test_recover_root_dir` like the directory for `$COMSPEC`, then it works.

justinmk commented:
> Expected 'rwx' but got 'rw-'
> Is this an upstream bug with libuv's us_fs_stat()?

I think pretty much every file in Windows is considered "executable". We fixed this for `executable()` (67fbbdb1b5eb6e48c7c5330) but did not address `getfperm()` specifically.

Also related is the TODO in #5229:

> os_setperm() needs to be re-thought; it is nonsense on Windows because of https://github.com/libuv/libuv/blob/e76b8838e51f6e8f1944b6c6d50c3948ed764a0b/src/win/fs.c#L1072-L1081

> Does neovim check if directory path in dir is valid such that it can't create a swapfile there and report E303?

It's likely that we're missing extra checks on Windows for that case.

justinmk commented on src/nvim/testdir/test_stat.vim:
```diff
@@ -39,11 +39,14 @@ func Test_existent_file()
 
   call assert_equal(0, getfsize(dname))
   call assert_equal('dir', getftype(dname))
-  call assert_equal(has('win32') ? 'rw-' : 'rwx', getfperm(dname)[0:2])
```
Let's keep this. I just looked more closely, and found that Vim's `getfperm` is inconsistent with `executable()`.

janlazo commented:
I added an explanation in https://github.com/neovim/neovim/pull/8160/commits/3a331251ce798445ee65de900741d2ef144ddbb9 for why `Test_recover_root_dir` fails in Appveyor. If the oldtests can run with reduced privs, then it will pass.

justinmk commented:
> If the oldtests can run with reduced privs, then it will pass.

Does that mean it will fail if run as Admin?

janlazo commented:
> Does that mean it will fail if run as Admin?

Yes. You can write in C:/Windows/System32 on Appveyor (idk if it's happens in the build phase only).

justinmk commented:
Is this at a stopping point?

janlazo commented:
Yes because of 2 blockers
1. UAC/ACL on Appveyor for user permissions
2. Reduce read/write privs for directories on Windows. Using `runas` to reduce privs on nvim works for files (ie. `:w /foo.txt` fails) but the swap directory is still created, even if empty, so `:split Xtest` doesn't fail for `Test_recover_root_dir`.

janlazo commented:
https://github.com/neovim/neovim/commit/afe6b4881f70df32107089d3c7df232d70eeda2e#diff-6a5a8a9f4d67827d3e91d79967aec3b2
This hack doesn't catch the retry error on the nodejs host and is not required for the mingw builds.

janlazo committed:
ci/win: don't use lastexitcode hack on mingw build

The heck is for MSVC builds to workaround msbuild error detection for cmake.

janlazo committed:
oldtests: win: keep set dir=/

Any user can create a directory on root.
Creating a directory on root allows any user to write files within that directory.

Test_recover_root_dir() passes when run locally as a regular user (not admin).

janlazo commented:
MINGW failure described in https://github.com/neovim/neovim/pull/8257#issuecomment-380712359

janlazo committed:
oldtests: win: skip test49

Vim doesn't test it on Windows.
See https://github.com/vim/vim/blob/master/src/testdir/Make_all.mak

b-r-o-c-k commented on ci/build.ps1:
```diff
@@ -97,15 +97,21 @@ bin\nvim --version ; exitIfFailed
 
 # Functional tests
 # The $LastExitCode from MSBuild can't be trusted
-$failed = $false
```
I don't think this is a problem because the MSBuild workaround should work with both MSVC and MinGW builds. The error message that it's searching for, `Running functional tests failed with error`, comes from `cmake/RunTests.cmake` and it should be printed when a test fails regardless of which generator is being used.

janlazo replied:
See the MINGW_32 log at https://ci.appveyor.com/project/neovim/neovim/build/19862/job/y452uy85m5fsl7rl
The build succeeded but it failed in the functional tests.

janlazo replied:
https://ci.appveyor.com/project/neovim/neovim/build/19866/job/2vpjs501eti96t5b
should have failed in functional tests but it ran the oldtests anyway.

b-r-o-c-k replied:
Yes, adding `2>&1` before the `|` should make it work. It appears that the error message isn't printed to stdout for the MinGW builds, so it wasn't being detected.

janlazo committed:
Revert "ci/win: don't use lastexitcode hack on mingw build"

This reverts commit bc43d2559f33aa6334d70834b389f0bc59dadcbf.

stderr redirection should make the workaround work for mingw builds.

janlazo committed:
ci/win: redirect stderr to fix mingw build

justinmk commented:
@janlazo Are you planning to open a new PR?

janlazo commented:
@justinmk No. I won't work on PRs until mid-May because of studies and personal matters. The work is done (could have left it as RDY and should be safe to merge as is) but I'm not interested in continuing further. Same applies to some PRs (ex. #8160) that I closed at the same time. I'm leaving my branches as is so other people can continue my work.

justinmk commented:
@janlazo Ok, thanks for your help thus far. I'll try to merge what I can.

blueyed commented:
> Do you know why `sourcing unix.vim` causes the oldtests to fail in Travis, specifically the filetype detection test?

This is due to "missing" `:filetype off` (just added to https://github.com/neovim/neovim/pull/10650).
